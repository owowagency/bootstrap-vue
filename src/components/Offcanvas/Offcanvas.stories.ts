import Btn from '@/components/Button';
import Offcanvas from '.';
import {placements} from '@/composables/usePlacement';

export default {
    title: 'Components/Offcanvas',
    component: Offcanvas,
    argTypes: {
        backdrop: {
            control: {type: 'boolean'},
        },
        header: {
            control: {type: 'text'},
        },
        noBody: {
            control: {type: 'boolean'},
        },
        placement: {
            control: {type: 'select'},
            options: placements,
        },
        scroll: {
            control: {type: 'boolean'},
        },
        show: {
            control: {type: 'boolean'},
        },
    },
};

const template = (args) => ({
    components: {Btn, Offcanvas},
    setup: () => ({args}),
    template: `
        <Btn
            data-bs-toggle="offcanvas"
            data-bs-target="#offcanvas"
        >
            Launch offcanvas
        </Btn>

        <p style="max-width: 100px;">
            Dummy content to make the container scrollable.
            Leave that to me. Send a distress signal, and inform the Senate that all on board were killed. Obi-Wan is here. The Force is with him. You mean it controls your actions? As you wish. I find your lack of faith disturbing.
            Hokey religions and ancient weapons are no match for a good blaster at your side, kid. Still, she's got a lot of spirit. I don't know, what do you think? Don't underestimate the Force. Still, she's got a lot of spirit. I don't know, what do you think?
        </p>

        <Offcanvas
            id="offcanvas"
            header="Offcanvas header"
            v-bind="args"
        >
            Offcanvas body
        </Offcanvas>
    `,
});

export const Default = template.bind({});
