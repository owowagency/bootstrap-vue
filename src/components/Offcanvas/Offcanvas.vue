<template>
    <div
        :id="id"
        ref="offcanvas"
        class="offcanvas"
        :class="classes"
        :data-bs-backdrop="backdrop"
        :data-bs-scroll="scroll"
    >
        <div
            v-if="!!$slots.header || !!header"
            class="offcanvas-header"
        >
            <!-- @slot Displays content inside the header of the offcanvas -->
            <slot name="header">
                {{ header }}
            </slot>
        </div>

        <div :class="{'offcanvas-body': !noBody}">
            <!-- @slot Displays the body of the offcanvas -->
            <slot />
        </div>
    </div>
</template>

<script lang="ts">
import {computed, ref} from 'vue';
import usePlacement, {placementProps} from '@/composables/usePlacement';
import {idProps} from '@/composables/useId';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';
import useClasses from '@/composables/useClasses';

const offcanvasEvents = ['show', 'shown', 'hide', 'hidden', 'hidePrevented'] as const;

type OffcanvasEvent = typeof offcanvasEvents[number];
</script>

<script lang="ts" setup>
const props = defineProps({
    /**
     * Defines if a backdrop should be shown
     */
    backdrop: {
        type: Boolean,
        default: true,
    },
    /**
     * The header text of the offcanvas
     */
    header: {
        type: String,
        default: undefined,
    },
    ...idProps,
    /**
     * Defines if the `offcanvas-body` class should be omitted
     */
    noBody: {
        type: Boolean,
        default: false,
    },
    ...placementProps,
    /**
     * Defines if the main content should be scrollable when offcanvas is opened
     */
    scroll: {
        type: Boolean,
        default: false,
    },
    /**
     * Defines if the offcanvas is shown
     */
    show: {
        type: Boolean,
        default: false,
    },
});

const emit = defineEmits<{(event: OffcanvasEvent): void}>();

const offcanvas = ref<HTMLElement>();

useBootstrapEmits(
    offcanvas,
    offcanvasEvents,
    emit,
    'offcanvas',
);

const {bsInstance: bsOffcanvas} = useBootstrapInstance(
    'Offcanvas',
    offcanvas,
);

const {classes} = useClasses(computed(() => [
    usePlacement(props.placement, 'offcanvas-{0}').placementClass.value,
    computed(() => props.show ? 'show' : undefined).value,
]));

defineExpose({bsOffcanvas});
</script>

<docs>
```vue
<Offcanvas header="Navigation">
    <Nav>
        <NavItem>Home</NavItem>
        <NavItem>Docs</NavItem>
        <NavItem>About</NavItem>
    </Nav>
</Offcanvas>
```

## Imported Props

| Prop name | Description                    | Type                                            | Values | Default |
| --------- | ------------------------------ | ----------------------------------------------- | ------ | ------- |
| id        | The id used for the component  | string                                          | -      | `uuid`  |
| placement | The placement of the component | [Size](../../composables/usePlacement) (string) | -      | 'start' |

<!-- TODO Auto-generate based on type? -->
## Events

| Name          | Args | Description                                                                                                               |
| ------------- | ---- | ------------------------------------------------------------------------------------------------------------------------- |
| show          |      | This event fires immediately when the `show` instance method is called                                                    |
| shown         |      | This event is fired when an offcanvas element has been made visible to the user                                           |
| hide          |      | This event is fired immediately when the `hide` instance method has been called                                           |
| hidden        |      | This event is fired when an offcanvas element has been hidden from the user                                               |
| hidePrevented |      | This event is fired when the offcanvas is shown, its backdrop is static and a click outside of the offcanvas is performed |
</docs>
