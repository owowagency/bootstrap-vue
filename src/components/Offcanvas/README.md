# Offcanvas

```vue
<Offcanvas header="Navigation">
    <Nav>
        <NavItem>Home</NavItem>
        <NavItem>Docs</NavItem>
        <NavItem>About</NavItem>
    </Nav>
</Offcanvas>
```

## Imported Props

| Prop name | Description                    | Type                                            | Values | Default |
| --------- | ------------------------------ | ----------------------------------------------- | ------ | ------- |
| id        | The id used for the component  | string                                          | -      | `uuid`  |
| placement | The placement of the component | [Size](../../composables/usePlacement) (string) | -      | 'start' |

<!-- TODO Auto-generate based on type? -->

## Events

| Name          | Args | Description                                                                                                               |
| ------------- | ---- | ------------------------------------------------------------------------------------------------------------------------- |
| show          |      | This event fires immediately when the `show` instance method is called                                                    |
| shown         |      | This event is fired when an offcanvas element has been made visible to the user                                           |
| hide          |      | This event is fired immediately when the `hide` instance method has been called                                           |
| hidden        |      | This event is fired when an offcanvas element has been hidden from the user                                               |
| hidePrevented |      | This event is fired when the offcanvas is shown, its backdrop is static and a click outside of the offcanvas is performed |

## Props

| Prop name | Description                                                               | Type    | Values | Default   |
| --------- | ------------------------------------------------------------------------- | ------- | ------ | --------- |
| backdrop  | Defines if a backdrop should be shown                                     | boolean | -      | true      |
| header    | The header text of the offcanvas                                          | string  | -      | undefined |
| noBody    | Defines if the `offcanvas-body` class should be omitted                   | boolean | -      | false     |
| scroll    | Defines if the main content should be scrollable when offcanvas is opened | boolean | -      | false     |
| show      | Defines if the offcanvas is shown                                         | boolean | -      | false     |

## Slots

| Name    | Description                                         | Bindings |
| ------- | --------------------------------------------------- | -------- |
| header  | Displays content inside the header of the offcanvas |          |
| default | Displays the body of the offcanvas                  |          |
