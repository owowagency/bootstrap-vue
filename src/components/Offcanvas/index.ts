import Offcanvas from './Offcanvas.vue';

export default Offcanvas;
