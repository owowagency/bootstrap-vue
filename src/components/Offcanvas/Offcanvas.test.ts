import Offcanvas from '.';
import {shallowMount} from '@vue/test-utils';

jest.mock('bootstrap', () => ({
    Offcanvas: {
        getOrCreateInstance: jest.fn(),
    },
}));

describe('template', () => {
    componentRenderTest(Offcanvas, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentSlotRenderTest(Offcanvas, 'header');

    componentSlotRenderTest(Offcanvas);

    componentWrapperAttributeTest(Offcanvas, {backdrop: true}, 'data-bs-backdrop', 'true');

    componentWrapperAttributeTest(Offcanvas, {scroll: true}, 'data-bs-scroll', 'true');

    componentWrapperClassTest(Offcanvas, {placement: 'end'}, 'offcanvas-end');

    componentWrapperClassTest(Offcanvas, {show: true}, 'show');

    it('renders header', () => {
        const text = 'I am a header';

        const wrapper = shallowMount(Offcanvas, {
            props: {header: text},
        });

        expect(wrapper.find('.offcanvas-header').text()).toBe(text);
    });

    it('does not add body class', () => {
        const wrapper = shallowMount(Offcanvas, {
            props: {noBody: true},
        });

        expect(wrapper.find('.offcanvas-body').exists()).toBe(false);
    });
});

describe('onMounted', () => {
    it('sets bsOffcanvas', async() => {
        const wrapper = await shallowMount(Offcanvas);

        expect((await import('bootstrap')).Offcanvas.getOrCreateInstance).toBeCalledWith(wrapper.vm.$refs.offcanvas, {});

        // TODO: Unable to assert bsOffcanvas value.
    });
});
