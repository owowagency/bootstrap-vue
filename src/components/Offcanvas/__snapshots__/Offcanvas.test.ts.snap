// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div id=\\"dont-change-please\\" class=\\"offcanvas offcanvas-start\\" data-bs-backdrop=\\"true\\" data-bs-scroll=\\"false\\">
  <!--v-if-->
  <div class=\\"offcanvas-body\\">
    <!-- @slot Displays the body of the offcanvas -->
  </div>
</div>"
`;
